test/Streams/SeekableReader.test.cpp

```diff
@@ -0,0 +1,65 @@
+#include "Streams/MemoryReader.h"
+#include "Streams/FileReader.h"
+#include "Streams/FileSliceReader.h"
+#include <gtest/gtest.h>
+#include <array>
+
+template <class T>
+Stream::SeekableReader* CreateSeekableReader();
+
+template <>
+Stream::SeekableReader* CreateSeekableReader<Stream::MemoryReader>() {
+	std::array<char, 5> buffer{ 't', 'e', 's', 't', '!' };
+	return new Stream::MemoryReader(buffer.data(), buffer.size() * sizeof(char));
+}
+
+template <>
+Stream::SeekableReader* CreateSeekableReader<Stream::FileReader>() {
+	return new Stream::FileReader("Streams/SimpleStream.txt");
+}
+
+template <>
+Stream::SeekableReader* CreateSeekableReader<Stream::FileSliceReader>() {
+	Stream::FileReader fileReader("Streams/SimpleStream.txt");
+	return &fileReader.Slice(5);
+}
+
+// Define a test fixture class template
+template <class T>
+class SimpleSeekableReader : public testing::Test {
+protected:
+	// The ctor calls the factory function to create a reader implemented by T
+	SimpleSeekableReader() : seekableReader(CreateSeekableReader<T>()) {}
+
+	virtual ~SimpleSeekableReader() { delete seekableReader; }
+
+	Stream::SeekableReader* const seekableReader;
+	const unsigned int size = 5;
```

Comment:
This is probably too detailed for a generic test fixture. It may not be feasible for some valid streams to be constructed of length 5.

One possible problem example is a stream which produces an infinite sequence of `0`s. It won't have a finite length. Though that might also make it not a seekable stream, in that it doesn't have a well defined `Length()` result.

A modified example is a stream that produces pages of `0`s, for some fixed size.

Another possible example, is a duplicator stream, which doubles everything read from it. It would necessarily have an even length.

Other odd cases, might be testing network sockets. You could do it, but you'd have to create both ends of the socket to feed it.

Reply:
It is hard to write these tests when we have different ideas on how the interface will be extended and the ideas are not documented anywhere. I thought a network socket would not be a SeekableReader but just a Stream::Reader.

I can separate this back out into three individual tests.

Reply:
You have a point, sockets generally aren't seekable. And yes, the other examples are a little bit esoteric to the point of perhaps not being useful. I suppose this is one of the reasons for testing though, in that writing good tests really forces you to tease out these little details.